src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+        {
+            var expression = await document.TryGetRelevantNodeAsync<TExpressionSyntax>(textSpan, cancellationToken).ConfigureAwait(false);
+            if (expression == null || CodeRefactoringHelpers.IsNodeUnderselected(expression, textSpan))
+            {
+                return null;
+            }
+
+            var semanticDocument = await SemanticDocument.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+
+            var expressionType = semanticDocument.SemanticModel.GetTypeInfo(expression, cancellationToken).Type;
+            if (expressionType is null or IErrorTypeSymbol)
+            {
+                return null;
+            }
+
+            var (title, actions) = AddActions(semanticDocument, expression, cancellationToken);
+
+            if (actions.Length > 0)
+            {
+                return new CodeActionWithNestedActions(title, actions, isInlinable: true);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Introduces a new parameter and refactors all the call sites
+        /// </summary>
+        public async Task<Solution> IntroduceParameterForRefactoringAsync(SemanticDocument document, TExpressionSyntax expression, bool allOccurrences, CancellationToken cancellationToken)
+        {
+            var parameterName = GetNewParameterName(document, expression, cancellationToken);
+            var methodSymbolInfo = GetMethodSymbolFromExpression(document, expression, cancellationToken)!;
+            var methodCallSites = await FindCallSitesAsync(document, methodSymbolInfo, cancellationToken).ConfigureAwait(false);
+
+            var updatedCallSitesSolution = await RewriteCallSitesAsync(document, expression, methodCallSites, allOccurrences, parameterName, cancellationToken).ConfigureAwait(false);
+            return updatedCallSitesSolution;
+        }
+
+        /// <summary>
+        /// Introduces a new parameter and a new method that calls upon the updated method so that all
+        /// refactorings are coalesced to one location
+        /// </summary>
+        public async Task<Solution> IntroduceParameterForTrampolineAsync(SemanticDocument document, TExpressionSyntax expression, bool allOccurrences, bool trampoline, CancellationToken cancellationToken)
+        {
+            var parameterName = GetNewParameterName(document, expression, cancellationToken);
+            var methodSymbolInfo = GetMethodSymbolFromExpression(document, expression, cancellationToken)!;
+            var methodCallSites = await FindCallSitesAsync(document, methodSymbolInfo, cancellationToken).ConfigureAwait(false);
+            var updatedInvocationsSolution = await RewriteCallSitesWithNewMethodAsync(document,
+                expression, methodSymbolInfo, allOccurrences, parameterName, methodCallSites, trampoline, cancellationToken).ConfigureAwait(false);
+            return updatedInvocationsSolution;
+        }
```

Comment:
all this code looks identical to the above method.  can we share, and just have them deviate on the trampoline case?

Reply:
yes, changed